Orange/widgets/utils/owlearnerwidget.py

```diff
@@ -164,23 +176,50 @@ def set_data(self, data):
                     "Select one with the Select Columns widget.")
             self.data = None
 
-        self.update_model()
+        # invalidate the model so that handleNewSignals will update it
+        self.model = None
+
 
     def apply(self):
+        level, self.__apply_level = max(self.__apply_level, default=self._UPDATE), []
         """Applies learner and sends new model."""
-        self.update_learner()
-        self.update_model()
+        if level == self._SEND:
+            self._send_learner()
+            self._send_model()
+        elif level == self._UPDATE:
+            self.update_learner()
+            self.update_model()
+        else:
+            self.learner or self.update_learner()
+            self.model or self.update_model()
+
+    def apply_as(self, level, unconditional=False):
+        self.__apply_level.append(level)
```

Comment:
This is horrendous, but I see no way around it (while maintaining compatibility).

This class and derived classes used `apply` to update and send both models; it would be better to update just the invalidated. Furthermore `_change_name` sent outputs by itself if `auto_apply` was `True`, but of course could not set the dirty flag if it was `False`. In some cases, this caused outputing the model twice...

To clean the mess, yet keep the compatilbility, I had a solution in which `apply` got an additional argument with default value. Unfortunatelly, `gui.auto_apply` doesn't pass extra arguments to `apply` -- and we shouldn't modify it to, just to fix this specific case.

To make it even more tricky, the function as defined here is called only when auto apply is on.

Hence, `apply_as` passes an argument by appending the "urgency" to the list. When the apply function is actually called, it finds the maximal "urgency" in the list and acts on it.